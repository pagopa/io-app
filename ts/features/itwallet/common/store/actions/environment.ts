import { ActionType, createStandardAction } from "typesafe-actions";
import { EnvType } from "../../utils/environment.ts";

export const itwSetEnv = createStandardAction("ITW_SET_ENV")<EnvType>();
export const itwResetEnv = createStandardAction("ITW_RESET_ENV")();

export type ItwEnvironmentActions =
  | ActionType<typeof itwSetEnv>
  | ActionType<typeof itwResetEnv>;
