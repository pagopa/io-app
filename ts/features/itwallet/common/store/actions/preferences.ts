import { ActionType, createStandardAction } from "typesafe-actions";

export const itwCloseFeedbackBanner = createStandardAction(
  "ITW_CLOSE_FEEDBACK_BANNER"
)();

export type ItwPreferencesActions = ActionType<typeof itwCloseFeedbackBanner>;
