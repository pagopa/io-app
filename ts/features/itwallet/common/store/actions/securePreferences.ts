import { ActionType, createStandardAction } from "typesafe-actions";

export const itwOfflineAccessCounterUp = createStandardAction(
  "ITW_OFFLINE_ACCESS_COUNTER_UP"
)();
export const itwOfflineAccessCounterReset = createStandardAction(
  "ITW_OFFLINE_ACCESS_COUNTER_RESET"
)();

export type ItwSecurePreferencesActions =
  | ActionType<typeof itwOfflineAccessCounterUp>
  | ActionType<typeof itwOfflineAccessCounterReset>;
