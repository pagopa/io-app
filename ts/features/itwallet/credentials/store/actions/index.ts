import { ActionType, createStandardAction } from "typesafe-actions";
import { StoredCredential } from "../../../common/utils/itwTypesUtils";

export const itwCredentialsStore = createStandardAction(
  "ITW_CREDENTIALS_STORE"
)<StoredCredential>();

export const itwCredentialsRemove = createStandardAction(
  "ITW_CREDENTIALS_REMOVE"
)<StoredCredential>();

export type ItwCredentialsActions =
  | ActionType<typeof itwCredentialsStore>
  | ActionType<typeof itwCredentialsRemove>;
