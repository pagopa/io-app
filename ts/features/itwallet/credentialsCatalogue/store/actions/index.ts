import { ActionType, createAsyncAction } from "typesafe-actions";
import { NetworkError } from "../../../../../utils/errors";
import { DigitalCredentialsCatalogue } from "../../../common/utils/itwCredentialsCatalogueUtils";

/**
 * This action handles the Digital Credentials Catalogue fetch
 */
export const itwFetchCredentialsCatalogue = createAsyncAction(
  "ITW_CREDENTIALS_CATALOGUE_REQUEST",
  "ITW_CREDENTIALS_CATALOGUE_SUCCESS",
  "ITW_CREDENTIALS_CATALOGUE_FAILURE"
)<void, DigitalCredentialsCatalogue, NetworkError>();

export type ItwCredentialsCatalogueActions = ActionType<
  typeof itwFetchCredentialsCatalogue
>;
