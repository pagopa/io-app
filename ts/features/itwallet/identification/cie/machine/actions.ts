import { CieManager } from "@pagopa/io-react-native-cie";
import { ActionArgs } from "xstate";
import I18n from "../../../../../i18n";
import {
  ItwFlow,
  trackItWalletCardReadingClose,
  trackItWalletCieCardReadingFailure,
  trackItWalletCieCardReadingSuccess,
  trackItWalletCieCardVerifyFailure,
  trackItWalletErrorPin,
  trackItWalletLastErrorPin,
  trackItWalletSecondErrorPin
} from "../../../analytics";
import { isNfcError } from "../utils/error";
import { CieContext } from "./context";
import { CieEvents } from "./events";

/**
 * Get the progress emojis based on the reading progress.
 * @param progress The reading progress value from 0 to 1.
 * @returns A string representing the progress bar with emojis,
 */
export const getProgressEmojis = (progress: number) => {
  // Clamp progress between 0 and 1
  const clampedProgress = Math.max(0, Math.min(1, progress));

  const totalDots = 8; // Length of the progress bar
  const blueDots = Math.floor(clampedProgress * totalDots);
  const whiteDots = totalDots - blueDots;

  const blueDotEmoji = "ðŸ”µ";
  const whiteDotyEmoji = "âšª";

  return blueDotEmoji.repeat(blueDots) + whiteDotyEmoji.repeat(whiteDots);
};

export const cieMachineActions = {
  /**
   * Configures the (idle) status alerts for the iOS NFC system dialog
   */
  configureStatusAlerts: () => {
    CieManager.setAlertMessage(
      "readingInstructions",
      I18n.t("features.itWallet.identification.cie.readingCard.ios.idle.status")
    );
    CieManager.setAlertMessage(
      "readingSuccess",
      I18n.t(
        "features.itWallet.identification.cie.readingCard.ios.completed.status"
      )
    );
  },

  /**
   * Updates the status alert for the iOS NFC system dialog with the current reading progress.
   */
  updateStatusAlert: ({
    context
  }: ActionArgs<CieContext, CieEvents, CieEvents>) => {
    const progress = getProgressEmojis(context.readProgress ?? 0);
    const label = I18n.t(
      "features.itWallet.identification.cie.readingCard.ios.reading.status"
    );
    CieManager.setCurrentAlertMessage(`${progress}\n${label}`);
  },

  trackSuccess: ({
    context: { isL3 }
  }: ActionArgs<CieContext, CieEvents, CieEvents>) => {
    trackItWalletCieCardReadingSuccess(isL3 ? "L3" : "L2");
  },

  trackError: ({
    context: { failure, isL3 }
  }: ActionArgs<CieContext, CieEvents, CieEvents>) => {
    const itw_flow: ItwFlow = isL3 ? "L3" : "L2";

    if (isNfcError(failure)) {
      switch (failure.name) {
        case "WEBVIEW_ERROR": // No tracking
          return;
        case "NOT_A_CIE":
          trackItWalletCieCardReadingFailure({
            reason: "unknown card",
            itw_flow
          });
          return;
        case "APDU_ERROR":
          trackItWalletCieCardReadingFailure({
            reason: failure.message,
            itw_flow
          });
          return;
        case "WRONG_PIN":
          if (failure.attemptsLeft > 1) {
            trackItWalletErrorPin(itw_flow);
          } else {
            trackItWalletSecondErrorPin(itw_flow);
          }
          return;
        case "CARD_BLOCKED":
          trackItWalletLastErrorPin(itw_flow);
          return;
        case "CERTIFICATE_EXPIRED":
        case "CERTIFICATE_REVOKED":
          trackItWalletCieCardVerifyFailure(itw_flow);
          return;
        case "CANCELLED_BY_USER":
          trackItWalletCardReadingClose();
          return;
      }
    }

    trackItWalletCieCardReadingFailure({ reason: "KO", itw_flow });
  }
};
