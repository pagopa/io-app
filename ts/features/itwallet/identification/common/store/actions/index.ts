import { ActionType, createAsyncAction } from "typesafe-actions";

export const itwHasNfcFeature = createAsyncAction(
  "ITW_HAS_NFC_FEATURE_REQUEST",
  "ITW_HAS_NFC_FEATURE_SUCCESS",
  "ITW_HAS_NFC_FEATURE_FAILURE"
)<void, boolean, Error>();

export type ItwIdentificationActions = ActionType<typeof itwHasNfcFeature>;
