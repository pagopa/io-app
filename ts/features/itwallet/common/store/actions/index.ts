import { ItwCredentialsActions } from "../../../credentials/store/actions";
import { ItwIdentificationActions } from "../../../identification/store/actions";
import { ItwIssuanceActions } from "../../../issuance/store/actions";
import { ItwLifecycleActions } from "../../../lifecycle/store/actions";

export type ItwActions =
  | ItwIdentificationActions
  | ItwIssuanceActions
  | ItwLifecycleActions
  | ItwCredentialsActions;
