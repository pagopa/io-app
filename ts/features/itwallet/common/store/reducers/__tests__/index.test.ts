import { applicationChangeState } from "../../../../../../store/actions/application";
import itWalletReducer from "../index";

describe("itWalletReducer", () => {
  it("should match snapshot", () => {
    const itWalletState = itWalletReducer(
      undefined,
      applicationChangeState("active")
    );
    expect(itWalletState).toMatchSnapshot();
  });
});
