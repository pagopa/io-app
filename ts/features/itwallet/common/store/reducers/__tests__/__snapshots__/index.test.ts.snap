// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`itWalletReducer should match snapshot 1`] = `
{
  "credentials": {
    "credentials": [],
    "eid": {
      "_tag": "None",
    },
  },
  "identification": {
    "isCieSupported": {
      "kind": "PotNone",
    },
    "isNfcEnabled": {
      "kind": "PotNone",
    },
  },
  "issuance": {
    "integrityKeyTag": {
      "_tag": "None",
    },
  },
  "lifecycle": 0,
  "preferences": {},
  "walletInstance": {
    "attestation": undefined,
  },
}
`;
